### PR TITLE
Gdb 6142 add failed transaction handling for kafka sink records

### PR DESCRIPTION
**Pull request by dobri-dimitrov**
Added Error logger and connection retry

### CODE REVIEW

File: src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java [hunk 1]
```diff
@@ -19,26 +27,69 @@
...
+				while (DEFERRED_TIME_BETWEEN_RETRIES > System.currentTimeMillis() - startAttempt) {
+					//do nothing
+				}
+				if (NUMBER_OF_CONNECTION_RETRIES <= retryCount) {
+					throw new TimeoutException("Flushing run out of time due to: " + e.getMessage());
+				}
+				LogErrorHandler.logInfo("Retrying connection...");
+				retryCount++;
 			}
+		} while (true);
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java:78]: never a good idea to have a while (true) block, why don't you use the retries count here

File: src/main/java/com/ontotext/kafka/service/UpdateRecordsProcessor.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import org.eclipse.rdf4j.rio.RDFFormat;
+
+import java.util.Collection;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class UpdateRecordsProcessor extends SinkRecordsProcessor {
+
+	// smart update template id
+
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/service/UpdateRecordsProcessor.java:15]: this is part of another task, remove from this MR

File: src/main/java/com/ontotext/kafka/util/PropertiesUtil.java [hunk 3]
```diff
@@ -16,20 +16,29 @@
...
+	public static final String DEFAULT_BOOTSTRAP_SERVERS = "localhost:9092,localhost:9093,localhost:9094";
+
 	private static final Logger LOGGER = LoggerFactory.getLogger(PropertiesUtil.class);
 
 	private static String version = "unknown";
 
 	private static Properties properties;
 
-	private PropertiesUtil() {}
+	private PropertiesUtil() {
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/util/PropertiesUtil.java:40]: revert unnecessary change

File: src/main/java/com/ontotext/kafka/util/PropertiesUtil.java [hunk 4]
```diff
@@ -16,20 +16,29 @@
...
 import org.slf4j.LoggerFactory;
 
+import java.util.Properties;
+
 public class PropertiesUtil {
 
+	public static final String CONNECTION_RETRY_DEFERRED_TIME = "connection.retry.deferred.time";
+	public static final long DEFAULT_CONNECTION_RETRY_DEFERRED_TIME = 100L;
+	public static final String CONNECTION_NUMBER_OF_RETRIES = "connection.retry.number.of.times";
+	public static final long DEFAULT_CONNECTION_NUMBER_OF_RETRIES = 300L;
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/util/PropertiesUtil.java:30]: these properties do not belong here, use config class

File: src/main/java/com/ontotext/kafka/util/PropertiesUtil.java [hunk 5]
```diff
@@ -16,20 +16,29 @@
...
+	public static final String BOOTSTRAP_SERVERS = "bootstrap.servers";
+	public static final String DEFAULT_BOOTSTRAP_SERVERS = "localhost:9092,localhost:9093,localhost:9094";
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/util/PropertiesUtil.java:32]: why would you need default localhost bootstrap servers?

File: src/main/java/com/ontotext/kafka/util/PropertiesUtil.java [hunk 6]
```diff
@@ -49,4 +58,22 @@ public static String getVersion() {
 	public static String getProperty(String key) {
 		return properties.getProperty(key);
 	}
+
+	public static long getFromPropertyOrDefault(String propertyName, Long defaultValue) {
+		String propertyValue = getProperty(propertyName);
+		try {
+			return Long.parseLong(propertyValue);
+
+		} catch (NumberFormatException e) {
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/util/PropertiesUtil.java:67]: these methods are unncessary

File: src/main/java/com/ontotext/kafka/error/LogErrorHandler.java [hunk 7]
```diff
@@ -1,12 +1,93 @@
...
+	}
+
+	public static void logWarning(String msg) {
+		LOGGER.warn(msg);
+	}
+
+	public static void logInfo(String msg) {
+		LOGGER.info(msg);
+	}
+
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/error/LogErrorHandler.java:48]: remove static methods, and use loggers that are defined for the right classes (e.g. addrecordsprocessor) as this will be misleading.

The sole purpose of this class is what the interface intended it to do. These static methods make the class not compliant with Liskov's substitution principle.

File: src/main/java/com/ontotext/kafka/error/LogErrorHandler.java [hunk 8]
```diff
@@ -1,12 +1,93 @@
...
+		void returnFailed(SinkRecord record) {
+			String recordKey = record.key() == null ? "null" : record.key().toString();
+			String recordValue = record.value() == null ? "null" : record.value().toString();
+			Producer<String, String> producer = new KafkaProducer<>(properties);
+
+			try {
+				ProducerRecord<String, String> pr = new ProducerRecord<>(ERRORS_DEAD_LETTER_QUEUE_TOPIC_NAME, recordKey, recordValue);
+				producer.send(pr, (metadata, exception) -> {
+					String failedRecordInfo = String.format("Record (key=%s value=%s)\n" +
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/error/LogErrorHandler.java:64]: Create a new ProducerErrorHandler that actually sends back the record. The purpose of the log error handler is to simply publish it in the log. That is why we have the ErrorHandler API and the records processor can be injected with either, depending on the selected config property. + add tests for it.

File: src/test/java/com/ontotext/kafka/error/LogErrorHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import org.junit.Test;
+
+public class LogErrorHandlerTest {
+
+	LogErrorHandler handler = new LogErrorHandler();
+
+/*	@Test
+	public void test() {
+		SinkRecord record = new SinkRecord("This is a failed record",
+				1,
```
Comment (tokovach) [inline src/test/java/com/ontotext/kafka/error/LogErrorHandlerTest.java:13]: do you plan on adding unit tests for the log error handler? If not, remove class.

File: src/test/java/com/ontotext/kafka/service/AddRecordsProcessorTest.java [hunk 10]
```diff
@@ -134,11 +135,187 @@ void testWriteBatchTimeoutAfterProperBatching() throws InterruptedException, IOE
...
+		}
+	}
+
+	@Test
+	@Disabled//it is slow
+	@DisplayName("Test should throw when out of retries")
+	@Timeout(35)
+	void testThrowTimoutExceptionException() {
+		int batch = 1;
+
```
Comment (tokovach) [inline src/test/java/com/ontotext/kafka/service/AddRecordsProcessorTest.java:272]: why is this test slow?

File: src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java [hunk 11]
```diff
@@ -19,26 +27,69 @@
...
+		}
+	}
+
+	private void flush() throws TimeoutException {
+		long retryCount = 1L;
+		Queue<SinkRecord> failedRecords = new ArrayDeque<>(recordsBatch.size());
+
+		do {
+			if (!failedRecords.isEmpty()) {
+				removeFailed(failedRecords);
```
Comment (tokovach) [inline src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java:54]: And if the batch size is 1 000 000, you would have to do linear searches every time. Remove failed records queue and handle each record directly by simply removing it and processing it with the error handler.

**Review by tokovach — CHANGES_REQUESTED** [review]

File: src/test/java/com/ontotext/kafka/service/AddRecordsProcessorTest.java (hunk 10 above)
Comment (dobri-dimitrov) [reply src/test/java/com/ontotext/kafka/service/AddRecordsProcessorTest.java:272]: because by default it will do 300 retries at rate 10/s ~ 30 sec

File: src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java (hunk 11 above)
Comment (dobri-dimitrov) [reply src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java:54]: IF you have 100 failed records and
IF you get an IOException at 999 999 record 
    than you will remove all failed 100 records from the Queue and retry the connection as long as it takes with all 999 **8**99 valid records.
